### PR TITLE
remove a bunch of ramda

### PR DESCRIPTION
**Pull request by hejkal**
**NOTE: DO NOT discuss internal CommitChange information in your PR; this PR will be public.
Link back to the issue in the Tix repo when you need to do that.**

### CODE REVIEW

File: client/js/campaigns/show/choose-gift-options-modal.js [hunk 1]
```diff
@@ -9,13 +8,15 @@ const giftOption = g => h('option', { props: {value: g.id} }, g.name)
...
-  state.selectedModalGift$(R.find(R.propEq('id', id))(gifts))
+  state.selectedModalGift$(gifts.find(i => i.id === id))
 }
 
 const chooseGift = (state, gifts) =>
   h('div.pastelBox--grey.u-padding--10', [
     h('select.u-margin--0', {on: {change: setDisplayGift(state, gifts)}}
-    , R.concat([h('option', 'Choose a gift option')], R.map(giftOption, gifts)))
+    , h('option', 'Choose a gift option')
+    , ...gifts.map(giftOption)
```
Comment (wwahammy) [inline client/js/campaigns/show/choose-gift-options-modal.js]: This supposed to be wrapped in an array? I'm not sure.

File: client/js/campaigns/show/metrics-and-contribute-box.js [hunk 2]
```diff
@@ -103,10 +102,16 @@ const progressBar = state => {
     current_status = (state.metrics$().starting_point || 0) + state.metrics$().total_raised;
   }
 
+  const clamp = (low, high, num) => {
```
Comment (wwahammy) [inline client/js/campaigns/show/metrics-and-contribute-box.js]: Is there a `clamp` in lodash?

File: client/js/campaigns/timeline.js [hunk 3]
```diff
@@ -23,8 +22,8 @@ function query() {
 }
 
 function cumulative(data) {
-  var moments = dateRange(R.head(data).date, R.last(data).date, 'days')
-  var dateStrings = R.map((m) => m.format('YYYY-MM-DD'), moments)
+  var moments = dateRange(data[0].date, data.slice(-1).date, 'days')
```
Comment (wwahammy) [inline client/js/campaigns/timeline.js]: Does this ever do anything different when data is empty?

File: client/js/common/class-object.js [hunk 4]
```diff
@@ -1,8 +1,7 @@
...
-
-module.exports = (classes='') => R.reduce(
-  (a, b) => {a[b] = true; return a}
-  , {}
-  , R.drop(1, classes.split('.')))
+module.exports = (classes = '') =>
+  classes.split('.').slice(1).reduce((a, b) => {
+    a[b] = true;
+    return a;
+  }, {});
```
Comment (wwahammy) [inline client/js/common/class-object.js]: If I correctly understand this, the code is taking string of css classes names and then creating an object of them. So if the person enters `'.foo.bar'` they're going to get an object like this:

```javascript
{
  foo: true,
  bar: true
 }
```

Is that how you understand it @hejkal?

File: client/js/common/request.js [hunk 5]
```diff
@@ -1,11 +1,11 @@
...
-const R = require('ramda')
 const request = require('flyd-ajax')
 
 module.exports = options => {
-  options.headers = R.merge({
-    'Content-Type': 'application/json'
-  , 'X-CSRF-Token': window._csrf
-  }, options.headers || {})
+  options.headers = {
+    ...options.headers,
```
Comment (wwahammy) [inline client/js/common/request.js]: Should this line be:

```...(options.headers || {})```

?

File: client/js/campaigns/timeline.js [hunk 6]
```diff
@@ -33,29 +32,29 @@ function cumulative(data) {
...
+  data.reduce((a, b) => {
     a[b.date] = b
     return a
-  }, dateDictionary, data)
+  }, dateDictionary)
   
-  return R.tail(R.reduce((a, b) => {
-    var last = R.last(a)
+  return Object.values(dateDictionary).reduce((a, b) => {
+    var last = a.slice(-1)
```
Comment (wwahammy) [inline client/js/campaigns/timeline.js]: Like before, is there any behavior different for corner cases when the array is empty?

File: client/js/recurring_donations/edit/index.es6 [hunk 7]
```diff
@@ -199,9 +197,9 @@ function view(state) {
         h('td.strong', 'Address')
       , h('td', [
           h('small', [
-            [supporter.address, supporter.city].filter(R.identity).join(', ')
+            [supporter.address, supporter.city].join(', ')
```
Comment (wwahammy) [inline client/js/recurring_donations/edit/index.es6]: What did `filter(R.identity)` do? What was trying to be achieved with it? Any ideas?

File: client/js/components/supporter-address-form.es6 [hunk 8]
```diff
@@ -15,11 +14,12 @@ const request = require('../common/request')
...
 function init(state) {
   state = state || {}
-  state = R.merge({
+  state = {
     submit$: flyd.stream()
   , supporter$: flyd.stream(state.supporter || {})
   , error$: flyd.stream()
-  }, state)
+  , ...state
+  }
```
Comment (wwahammy) [inline client/js/components/supporter-address-form.es6:23]: Is this the correct order? Do the properties of state go first or second?

File: client/js/components/supporter-address-form.es6 [hunk 9]
```diff
@@ -81,7 +81,10 @@ function view(state) {
       ])
     ])
   , h('input', {props: {type: 'hidden', name: 'id', value: supporter.id}})
-  , button(R.pick(['loading$', 'error$'], state))
+  , button({
+    ...(state.loading$ && { loading$: state.loading$ }),
+    ...(state.error$ && { error$: state.error$ }),
```
Comment (wwahammy) [inline client/js/components/supporter-address-form.es6]: Does this work? I can't get it to work in the console.

Alternatively, feel free to use a function from lodash if it's easier.

File: client/js/recurring_donations/edit/index.es6 [hunk 10]
```diff
@@ -363,7 +361,10 @@ const paydateForm = state =>
       }
     })
   , h('br')
-  , button(R.pick(['loading$', 'error$'], state))
+  , button({
+    ...(state.loading$ && { loading$: state.loading$ }),
+    ...(state.error$ && { error$: state.error$ }),
```
Comment (wwahammy) [inline client/js/recurring_donations/edit/index.es6]: This is another place that I'm unsure on whether this is correct. When a boolean is false, what happens?

File: client/js/common/direct-to-s3-upload.es6 [hunk 11]
```diff
@@ -28,7 +28,7 @@ const uploadFile = R.curry(input => {
       var url = `${presignedPost.s3_direct_url}`
       var file = input.files[0]
       var fileUrl = `${url}/tmp/${presignedPost.s3_uuid}/${file.name}`
-      var payload = R.merge(JSON.parse(presignedPost.s3_presigned_post), {file})
+      var payload = { ...JSON.parse(presignedPost.s3_presigned_post), ...{file} }
```
Comment (wwahammy) [inline client/js/common/direct-to-s3-upload.es6]: Once again, which object's properties go first and which ones should go second?

**Review by wwahammy — COMMENTED** [review]
Looking good! I'm looking forward to getting rid of ramda.

File: client/js/components/supporter-address-form.es6 (hunk 9 above)
Comment (hejkal) [reply client/js/components/supporter-address-form.es6]: it should work if I did it right, it's a pattern I've used elsewhere for inline optional object properties, but I can look at lodash

File: client/js/common/direct-to-s3-upload.es6 (hunk 11 above)
Comment (hejkal) [reply client/js/common/direct-to-s3-upload.es6]: As far as I can tell from the docs (which don't go back to the version this is on) the second one is merged into the first / overwrites values on it
https://ramdajs.com/0.25.0/docs/#merge

this could be simplified more to this though:
```js
var payload = { ...JSON.parse(presignedPost.s3_presigned_post), file }
```

File: client/js/components/supporter-address-form.es6 (hunk 9 above)
Comment (wwahammy) [reply client/js/components/supporter-address-form.es6]: Interesting. I'm trying to do it in the javascript console in Firefox and it doesn't work but I'm not sure if I'm missing something.

File: client/js/campaigns/show/choose-gift-options-modal.js (hunk 1 above)
Comment (hejkal) [reply client/js/campaigns/show/choose-gift-options-modal.js]: good catch, I think you're right — the R.concat would've been putting that all into an array.

File: client/js/campaigns/show/metrics-and-contribute-box.js (hunk 2 above)
Comment (hejkal) [reply client/js/campaigns/show/metrics-and-contribute-box.js]: sure is, good idea, I'll update

File: client/js/common/class-object.js (hunk 4 above)
Comment (hejkal) [reply client/js/common/class-object.js]: that sounds right to me, yeah. and I guess it's dropping / splitting out the first one because it's a string with leading `.`s? Like '.name.another-name'?

I dug into this a little more, and `class-object` is only being used by `number-input.js`, `text-input.js`, and `textarea.js`. Of those, only `text-input` is used anywhere else, and it's never passed a `classes` string. So I don't think this (or `number-input` or `textarea`) are used. Should I delete them instead of updating?

File: client/js/common/request.js (hunk 5 above)
Comment (hejkal) [reply client/js/common/request.js]: oops, this should be in the other order, and yes, I'll fix it

File: client/js/components/supporter-address-form.es6 (hunk 8 above)
Comment (hejkal) [reply client/js/components/supporter-address-form.es6:23]: looks from the docs like the second should override the first, so this order should work the same, yeah
https://ramdajs.com/0.25.0/docs/#merge

File: client/js/components/supporter-address-form.es6 (hunk 9 above)
Comment (hejkal) [reply client/js/components/supporter-address-form.es6]: Huh, this may have been where I found it first, not certain though:
https://medium.com/@slamflipstrom/conditional-object-properties-using-spread-in-javascript-714e0a12f496

I do see some people saying it's safer to use a ternary with null than && though

Comment (hejkal) [reply client/js/components/supporter-address-form.es6]: It's working for me in the console

Screenshot 2024-12-13 at 5 17 28 PM

File: client/js/recurring_donations/edit/index.es6 (hunk 7 above)
Comment (hejkal) [reply client/js/recurring_donations/edit/index.es6]: Ahh I see, it is doing something, good catch. If you filter by the thing itself, it gets past the filter if it's truthy, but not if it's falsy. So you get things like
```js
[1, 2, false, 3, 0, 4, ''].filter(i => i)  // [1, 2, 3, 4]
```
so I should put this back

Comment (hejkal) [reply client/js/recurring_donations/edit/index.es6]: (and the one just below)

File: client/js/components/supporter-address-form.es6 (hunk 9 above)
Comment (hejkal) [reply client/js/components/supporter-address-form.es6]: lodash has `_.pick` though, so I can just use that, it's definitely more concise for these situations

File: client/js/recurring_donations/edit/index.es6 (hunk 10 above)
Comment (hejkal) [reply client/js/recurring_donations/edit/index.es6]: it should work, commented on the other one, but as with there, I'll just use `pick` from lodash, it's more readable.

File: client/js/common/class-object.js (hunk 4 above)
Comment (wwahammy) [reply client/js/common/class-object.js]: Ya, I'd delete. Please do in a different PR.

File: client/js/campaigns/show/page.js [hunk 12]
```diff
@@ -34,7 +32,6 @@ if(app.current_campaign_editor) {
 	require('./admin')
 	appl.def('current_campaign_editor', true)
 	require('../../gift_options/admin')
-	var create_info_card = require('../../supporters/info-card.es6')
```
Comment (wwahammy) [inline client/js/campaigns/show/page.js]: Is this related to the ramda changes? If not, please do in a separate PR.

File: client/js/campaigns/show/page.js [hunk 13]
```diff
@@ -1,14 +1,12 @@
...
 const flyd = require('flyd')
 const h = require('snabbdom/h')
-const R = require('ramda')
 const donateWiz = require('../../nonprofits/donate/wizard')
 const render = require('ff-core/render')
 const snabbdom = require('snabbdom')
 const modal = require('ff-core/modal')
 flyd.mergeAll = require('flyd/module/mergeall')
 flyd.scanMerge = require('flyd/module/scanmerge')
-const format = require('../../common/format')
```
Comment (wwahammy) [inline client/js/campaigns/show/page.js]: Is this related to the ramda changes? If not, please do in a separate PR.

File: client/js/recurring_donations/edit/index.es6 [hunk 14]
```diff
@@ -199,9 +198,9 @@ function view(state) {
         h('td.strong', 'Address')
       , h('td', [
           h('small', [
-            [supporter.address, supporter.city].filter(R.identity).join(', ')
+            [supporter.address, supporter.city].filter(i => i).join(', ')
```
Comment (wwahammy) [inline client/js/recurring_donations/edit/index.es6:202]: Is this filter here to strip out elements that are empty?

File: client/js/recurring_donations/edit/index.es6 [hunk 15]
```diff
@@ -349,22 +348,22 @@ const updateAddressModal = state =>
...
+        type: 'number',
+        max: 28,
+        min: 1,
+        name: 'paydate',
+        value: state.recDon$().paydate || 1,
+      },
+    }),
+    h('br'),
+    button(pick(state, ['loading$', 'error$'])),
+  ]);
```
Comment (wwahammy) [inline client/js/recurring_donations/edit/index.es6]: Would it be possible to revert all of these changes except the `pick` change?

File: client/js/settings/index/email-settings/index.js [hunk 16]
```diff
@@ -1,9 +1,7 @@
 // License: LGPL-3.0-or-later
 // npm
-const snabbdom = require('snabbdom')
 const flyd = require('flyd')
 const R = require('ramda')
-const render = require('ff-core/render')
```
Comment (wwahammy) [inline client/js/settings/index/email-settings/index.js]: Are these related to the ramda changes? If not, let's do it in a separate PR.

File: client/js/supporters/settings/mailchimp-integration-settings.js [hunk 17]
```diff
@@ -14,11 +13,11 @@ function init(modalID$) {
...
   var state = {
     submitForm$: flyd.stream()
-  , tagMasters$: flyd.map(R.prop('body'), request({method: 'get', path: pathPrefix + '/tag_masters'}).load)
+  , tagMasters$: flyd.map(r => r.body, request({method: 'get', path: pathPrefix + '/tag_masters'}).load)
   }
 
-  const emailLists$ = flyd.map(R.prop('body'), request({method: 'get', path: pathPrefix + '/email_lists'}).load)
-  state.selectedTagMasterIds$ = flyd.map(R.map(ls => ls.tag_master_id), emailLists$)
+  const emailLists$ = flyd.map(r => r.body, request({method: 'get', path: pathPrefix + '/email_lists'}).load)
+  state.selectedTagMasterIds$ = flyd.map(emailLists$.map(ls => ls.tag_master_id))
```
Comment (wwahammy) [inline client/js/supporters/settings/mailchimp-integration-settings.js:20]: This one is wrong, I think.

File: client/js/recurring_donations/edit/get-params.js [hunk 18]
```diff
@@ -1,23 +1,21 @@
...
-  // Preprocess data
-  const evolve = R.evolve({
-    multiple_designations: splitParam
-  , custom_amounts: amts => R.compose(R.map(Number), splitParam)((amts instanceof String ? amts : R.map(x => x/100, amts).join(',')) || defaultAmts)
-  , custom_fields: fields => R.map(f => {
-      const [name, label] = R.map(R.trim, R.split(':', f))
-      return {name, label: label ? label : name}
-    }, R.split(',',  fields))
-  })
-  return R.compose(evolve, merge)(params)
```
Comment (wwahammy) [inline client/js/recurring_donations/edit/get-params.js]: So, this file is really doing (almost) the same thing as `client/js/nonprofits/donate/get-params.js`. I'd prefer to revert this change here and then work on combining the two together.

File: client/js/recurring_donations/edit/change-amount-wizard.es6 [hunk 19]
```diff
@@ -3,7 +3,6 @@ const flyd = require('flyd')
 const R = require('ramda')
 const h = require('snabbdom/h')
 const url = require('url')
-const render = require('ff-core/render')
```
Comment (wwahammy) [inline client/js/recurring_donations/edit/change-amount-wizard.es6]: Is this related to removing ramda? If not, let's do as separate PR.

**Review by wwahammy — CHANGES_REQUESTED** [review]
So far, this is great work @hejkal, thank you so much!

I've got some more to look through but I think this PR is actually too big already. The lack of front end and integration tests makes this really hard to guarantee we're not breaking anything. It's obviously good practice in general but given that it's open source, we have a secondary reason: if I (or theoretically anyone) wants to upstream any of the work, it becomes harder to manage cherry-picking. As I often say, doing things in the open incentivizes good engineering practices :smile: 


All that said, outside of a few things I've mentioned, I don't think it make sense to do go back and separate this all. Let's wrap up the work that's already been done and do additional changes in separate, more easily digestable and potentially reusable PRs.

File: client/js/recurring_donations/edit/change-amount-wizard.es6 (hunk 19 above)
Comment (hejkal) [reply client/js/recurring_donations/edit/change-amount-wizard.es6]: @wwahammy I can split these out yep. There are a few of them so far I think, just removing unused import lines as I come across them. So cleanup related, but not ramda-specific.

File: client/js/common/class-object.js (hunk 4 above)
Comment (hejkal) [reply client/js/common/class-object.js]: Sounds good, will do

File: client/js/recurring_donations/edit/index.es6 (hunk 14 above)
Comment (hejkal) [reply client/js/recurring_donations/edit/index.es6:202]: I assume so, I'm not sure what else its purpose would be

File: client/js/recurring_donations/edit/index.es6 (hunk 15 above)
Comment (hejkal) [reply client/js/recurring_donations/edit/index.es6]: yep!

File: client/js/recurring_donations/edit/get-params.js (hunk 18 above)
Comment (hejkal) [reply client/js/recurring_donations/edit/get-params.js]: sounds good, reverted

File: client/js/common/class-object.js (hunk 4 above)
Comment (hejkal) [reply client/js/common/class-object.js]: https://github.com/CommitChange/houdini/pull/966

File: client/js/supporters/settings/mailchimp-integration-settings.js (hunk 17 above)
Comment (hejkal) [reply client/js/supporters/settings/mailchimp-integration-settings.js:20]: Can you say more about what looks wrong here? As far as I can tell it's the same as other similar places, switching from `R.prop('thing')` to `i => i.thing` and from `R.map(f, thing)` to `thing.map(f)`

File: client/js/recurring_donations/edit/index.es6 (hunk 14 above)
Comment (hejkal) [reply client/js/recurring_donations/edit/index.es6:202]: But I can try to investigate further what happens if one is omitted/empty

File: client/js/campaigns/timeline.js (hunk 3 above)
Comment (hejkal) [reply client/js/campaigns/timeline.js]: `R.head` and `[][0]` are both `undefined` as far as I can tell. But good catch, it looks like `R.last` returns `undefined` where `[].slice(-1)` returns `[]`.  So we should maybe do `data[data.length - 1]` instead.

File: client/js/campaigns/timeline.js (hunk 6 above)
Comment (hejkal) [reply client/js/campaigns/timeline.js]: Yep, I think similarly here we should use `a[a.length-1]` (I wish there was cleaner JS for this)

**Comment by hejkal** [comment]
closing this because I split it out into smaller chunks
